wolfcrypt/src/sha3.c

```diff
@@ -569,8 +569,12 @@ static word64 Load64BitBigEndian(const byte* a)
     n |= ((word64)*(word16*)(a + 6)) << 48;
 
     return n;
-#else
+#elif ((WOLFSSL_GENERAL_ALIGNMENT > 0) && (WOLFSSL_GENERAL_ALIGNMENT == 8))
     return *(const word64*)a;
+#else
+    word64 n;
```

Comment:
Should this be conditional on `XSTREAM_ALIGN`?

Reply:
XSTREAM_ALIGN informs WOLFSSL_GENERAL_ALIGNMENT.

Reply:
WOLFSSL_GENERAL_ALIGNMENT may be 0 which I take to mean: "no alignment issues".

Reply:
Ahh I see. With `XSTREAM_ALIGN` I get ed448 sign and verify reports. Should the alignment of data be handled be the caller or in `ed448.c`?

<details>
<summary>ed448_verify_msg</summary>

```
wolfcrypt/src/sha3.c:559:19: runtime error: load of misaligned address 0x5605ddd4e9b9 for type 'word3
2', which requires 4 byte alignment
0x5605ddd4e9b9: note: pointer points here
 02 02 01  01 30 05 06 03 2b 65 71  30 81 99 31 0b 30 09 06  03 55 04 06 13 02 55 53  31 10 30 0e 06
              ^ 
    #0 0x7f6d75398ecc in Load64BitBigEndian wolfcrypt/src/sha3.c:559
    #1 0x7f6d75398ecc in Sha3Update wolfcrypt/src/sha3.c:635
    #2 0x7f6d755642c4 in ed448_verify_msg wolfcrypt/src/ed448.c:413
    #3 0x7f6d755654f8 in wc_ed448_verify_msg wolfcrypt/src/ed448.c:467
    #4 0x7f6d753e7f30 in ConfirmSignature wolfcrypt/src/asn.c:7926
    #5 0x7f6d753ea1fd in ParseCertRelative wolfcrypt/src/asn.c:10389
    #6 0x7f6d755dd03c in ProcessPeerCertParse src/internal.c:11094
    #7 0x7f6d7565a886 in ProcessPeerCerts src/internal.c:11514
    #8 0x7f6d7566152f in DoCertificate src/internal.c:12500
    #9 0x7f6d7566152f in DoHandShakeMsgType src/internal.c:13415
    #10 0x7f6d756707e4 in DoHandShakeMsg src/internal.c:13627
    #11 0x7f6d756707e4 in ProcessReply src/internal.c:16333
    #12 0x7f6d756b92ee in wolfSSL_connect src/ssl.c:12744
    #13 0x5605dc13620a in client_test examples/client/client.c:3232
    #14 0x5605dc123db4 in execute_test_case tests/suites.c:551
    #15 0x5605dc1250cf in test_harness tests/suites.c:727
    #16 0x5605dc125d63 in SuiteTest tests/suites.c:905
    #17 0x5605dc057161 in unit_test tests/unit.c:85
    #18 0x5605dc057161 in unit_test tests/unit.c:46
    #19 0x7f6d7445a0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

```

</details>

<details>
<summary>ed448_sign_msg</summary>

```
wolfcrypt/src/sha3.c:560:19: runtime error: load of misaligned address 0x7ffd5183f609 for type 'word32', which requires 4 byte alignment
0x7ffd5183f609: note: pointer points here
 0c 41 e6  a4 0c 93 1d b2 7b 2f aa  15 a8 ce dd 52 cf f7 36  2c 4e 6e 23 da ec 0f bc  3a 79 b6 80 6e
              ^ 
    #0 0x7f318cf97fd5 in Load64BitBigEndian wolfcrypt/src/sha3.c:560
    #1 0x7f318cf97fd5 in Sha3Update wolfcrypt/src/sha3.c:635
    #2 0x7f318d162e19 in ed448_sign_msg wolfcrypt/src/ed448.c:207
    #3 0x7f318d164417 in wc_ed448_sign_msg wolfcrypt/src/ed448.c:280
    #4 0x5571e81f7931 in ed448_test wolfcrypt/test/test.c:26838
    #5 0x5571e81ff055 in wolfcrypt_test wolfcrypt/test/test.c:1275
    #6 0x5571e821947b in testsuite_test testsuite/testsuite.c:153
    #7 0x7f318c0590b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #8 0x5571e81ae02d in _start (/home/elms/work/wolf/wolfssl/testsuite/.libs/testsuite.test+0x3b02d)

```

</details>

Reply:
What compiler are you using?
I can't get reports with gcc9 or clang10 on my Linux box.

Reply:
I used gcc 9.3.0 on a linux box.

```
./configure CFLAGS=-fsanitize=undefined\ -g --enable-all
UBSAN_OPTIONS=print_stacktrace=1:log_path=./UBSAN make test
```

This generates the two logs `UBSAN.<pid>` where pid was the pid of the process for testsuite and unit.test. I used `-g` to get the lines in the tracebacks.

Reply:
Able to reproduce now.
Change the loading of the data (sha3.c:599) to call Load64Unaligned():
```
WC_STATIC WC_INLINE word64 Load64Unaligned(unsigned char *a)
{
    return ((word64)a[0] <<  0) | 
           ((word64)a[1] <<  8) | 
           ((word64)a[2] << 16) | 
           ((word64)a[3] << 24) | 
           ((word64)a[4] << 32) | 
           ((word64)a[5] << 40) | 
           ((word64)a[6] << 48) | 
           ((word64)a[7] << 56);
}
```
gcc converts this to a single instruction.

Reply:
Thanks, the line number didn't match but think I changed it as you were suggesting.

Performance on my machine seem to have negligible impact. Are there specific configs that may impact this more?

<details>
<summary>head of tree</summary>

```
$ ./config.status --c
'--enable-all'
$ git describe 
v4.7.1r-599-gae2f2b246
$ ./wolfcrypt/benchmark/benchmark -csv -rabbit -hc128 -sha -sha2 -sha224 -sha256 -sha384 -sha512 -sha3
------------------------------------------------------------------------------
 wolfSSL version 4.7.1
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
This format allows you to easily copy the output to a csv file.

Symmetric Ciphers:

Algorithm,MB/s,Cycles per byte,
HC128,1712.890,1.89,
RABBIT,700.209,4.63,
SHA,513.502,6.31,
SHA-224,217.334,14.92,
SHA-256,220.317,14.71,
SHA-384,336.544,9.63,
SHA-512,338.646,9.57,
SHA3-224,307.666,10.54,
SHA3-256,291.198,11.13,
SHA3-384,220.581,14.70,
SHA3-512,154.931,20.92,
Benchmark complete

```

</details>

<details>
<summary>This PR</summary>

```
$ ./config.status --c
'--enable-all'
$ git describe 
v4.7.1r-601-gbcee1dbf6
$ ./wolfcrypt/benchmark/benchmark -csv -rabbit -hc128 -sha -sha2 -sha224 -sha256 -sha384 -sha512 -sha3
------------------------------------------------------------------------------
 wolfSSL version 4.7.1
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
This format allows you to easily copy the output to a csv file.

Symmetric Ciphers:

Algorithm,MB/s,Cycles per byte,
HC128,1681.548,1.93,
RABBIT,712.736,4.55,
SHA,515.687,6.29,
SHA-224,221.840,14.61,
SHA-256,221.584,14.63,
SHA-384,337.009,9.62,
SHA-512,339.915,9.54,
SHA3-224,309.002,10.49,
SHA3-256,291.040,11.14,
SHA3-384,224.874,14.41,
SHA3-512,154.027,21.05,
Benchmark complete
```


</details>

Reply:
No other configs. Just wanted you to check against master.
Looks fine.